opt: index constraint generation code for simple expressions

RaduBerinde opened the pull request:
#### opt: collapse multiple AND/OR nodes

This commit introduces a `normalizeScalar` routine that implements coalescing of
multiple binary AND/OR expressions into a single k-ary AND/OR expression. This
is convenient for the index constraints generation code.

Note that in the short term, because we use a `TypedExpr` that went through a
constant-folding and normalization process, we don't implement those
normalization rules here. Longer-term however, we probably want to move all
normalization to this package.

Release note: None

#### opt: index constraint generation code for simple expressions

This commit is the beginning of the new index constraint generation
code.

Currently we only support conjunctions of simple comparison
operations. However, the code was designed with more complex
operations in mind.

The `LogicalSpan` structure is a replacement for `sql.logicalSpan`; it
is a more efficient representation of the same concept: we don't need to store
an "inclusive" flag for every column; only the last column constraint can be
exclusive; it also makes more sense conceptually to refer to the entire key as
exclusive vs inclusive.

Constraints are represented as slices of disjoint, ordered
`LogcicalSpans`.

The general approach to calculating the constraints is different than
`sql.makeIndexConstraints`. That code starts from the front (first
index column), generates constraints, then tries to extend them. This
leads to very tricky situations when the start and end constraints
don't cover the same prefix of the index columns. In addition, during
this process, the constraints are not represented as logical spans but
as start and end expressions, which also leads to difficulties. The
new code (described in calcDepth) uses a back-to-front approach where
we compute `LogicalSpans` for each depth separately.

Currently, the code assumes all columns are ascending. Support for
directions will be added later.

Updates #6346.

CC @knz @jordanlewis

RaduBerinde committed:
opt: collapse multiple AND/OR nodes

This commit introduces a `normalizeScalar` routine that implements coalescing of
multiple binary AND/OR expressions into a single k-ary AND/OR expression. This
is convenient for the index constraints generation code.

Note that in the short term, because we use a `TypedExpr` that went through a
constant-folding and normalization process, we don't implement those
normalization rules here. Longer-term however, we probably want to move all
normalization to this package.

Release note: None

RaduBerinde committed:
opt: index constraint generation code for simple expressions

This commit is the beginning of the new index constraint generation
code.

Currently we only support conjunctions of simple comparison
operations. However, the code was designed with more complex
operations in mind.

The `LogicalSpan` structure is a replacement for `sql.logicalSpan`; it
is a more efficient representation of the same concept: we don't need to store
an "inclusive" flag for every column; only the last column constraint can be
exclusive; it also makes more sense conceptually to refer to the entire key as
exclusive vs inclusive.

Constraints are represented as slices of disjoint, ordered
`LogcicalSpans`.

The general approach to calculating the constraints is different than
`sql.makeIndexConstraints`. That code starts from the front (first
index column), generates constraints, then tries to extend them. This
leads to very tricky situations when the start and end constraints
don't cover the same prefix of the index columns. In addition, during
this process, the constraints are not represented as logical spans but
as start and end expressions, which also leads to difficulties. The
new code (described in calcDepth) uses a back-to-front approach where
we compute `LogicalSpans` for each depth separately.

Currently, the code assumes all columns are ascending. Support for
directions will be added later.

Updates #6346.

RaduBerinde commented:
Corrected the commit message to mention we don't support column directions yet.

petermattis commented:
Excited to see this progress. Just a few comments in a first pass. I'll have more later.

---

Reviewed 4 of 4 files at r1.
Review status: 3 of 7 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt/expr.go, line 58 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-L0BXnmj1XuMm1B7dfDx:-L0BXnmj1XuMm1B7dfDy:bjcvfcq) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad13f52973ad42906a2a283619d0eb25ac79de43/pkg/sql/opt/expr.go#L58)):*
> ```Go
> 		n := tp.Child(title)
> 		for _, e := range exprs {
> 			if e != nil {
> ```

FYI, I believe this happens for expressions extracted from the memo in some case. I can't remember the scenario, but if you comment the similar code out in `opttoy` you can see what breaks.

---

*[pkg/sql/opt/index_constraints.go, line 85 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-L0BZGDbUi1rj9nvz3wq:-L0BZGDbUi1rj9nvz3wr:baeo776) ([raw file](https://github.com/cockroachdb/cockroach/blob/16bf712e73eb4df95ef8ce4842c61a34ddbc26c1/pkg/sql/opt/index_constraints.go#L85)):*
> ```Go
> }
> 
> // Sting formats a LogicalSpan. Inclusivity/exclusivity is shown using
> ```

`s/Sting/String/g`

---

*[pkg/sql/opt/index_constraints.go, line 189 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-L0BZxZKvhZQ9rhd0T5R:-L0BZxZKvhZQ9rhd0T5S:bwek32f) ([raw file](https://github.com/cockroachdb/cockroach/blob/16bf712e73eb4df95ef8ce4842c61a34ddbc26c1/pkg/sql/opt/index_constraints.go#L189)):*
> ```Go
> // If we are comparing start keys, direction must be +1.
> // If we are comparing end keys, direction must be -1.
> func (c *indexConstraintCalc) compare(a, b LogicalKey, direction int) int {
> ```

Consider adding an enum for `direction` to enforce a `+1` or `-1` value.

---

*[pkg/sql/opt/index_constraints.go, line 331 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-L0Bedhv3Xc8Jc2IRoMb:-L0Bedhv3Xc8Jc2IRoMc:b8qk43b) ([raw file](https://github.com/cockroachdb/cockroach/blob/16bf712e73eb4df95ef8ce4842c61a34ddbc26c1/pkg/sql/opt/index_constraints.go#L331)):*
> ```Go
> 	switch e.op {
> 	case eqOp, ltOp, gtOp, leOp, geOp:
> 		if !(e.children[0].op == variableOp && e.children[0].private.(int) == depth) {
> ```

I don't think we want to leak the contents of `private` throughout the code. Perhaps pull this out into an `isIndexedVar` method. Or should we be populating something in `scalarProps` with regards to variable indexes? I guess that might be a TODO for the future given that this code doesn't have variable indexes yet. Seems like `indexConstraintCalc` will also need a slice of indexes for the variables composing the index.

---

*[pkg/sql/opt/scalar.go, line 143 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-L0BY3EgL-VN2Lo4wS5p:-L0BY3EgL-VN2Lo4wS5q:b57h2iz) ([raw file](https://github.com/cockroachdb/cockroach/blob/159c322300387675b7e70767f610f05644388a38/pkg/sql/opt/scalar.go#L143)):*
> ```Go
> // Applies a set of normalization rules to a scalar expression.
> //
> // For now, we expect to build exprs from TypedExprs which have went through a
> ```

`s/have went/have gone/g`

---

*[pkg/sql/opt/testdata/build-scalar, line 48 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-L0BYI1hukaHujcksCbf:-L0BYI1iOKHfBf45W6qC:b-79ioxf) ([raw file](https://github.com/cockroachdb/cockroach/blob/159c322300387675b7e70767f610f05644388a38/pkg/sql/opt/testdata/build-scalar#L48)):*
> ```
> 
> build-scalar,normalize int int
> @1 >= 5 AND @1 <= 10 AND @2 > 0 AND @2 < 10
> ```

I'd like to see some other test cases here, such as normalize ORs and various combinations of ANDs and ORs.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-:-L0BgCb_FW8_2qaDe012:bfr4jx1)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 1 of 7 files reviewed at latest revision, 6 unresolved discussions.

---

*[pkg/sql/opt/expr.go, line 58 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-L0BXnmj1XuMm1B7dfDx:-L0BlfYNIgeCa041sjmU:b-290j9v) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad13f52973ad42906a2a283619d0eb25ac79de43/pkg/sql/opt/expr.go#L58)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

FYI, I believe this happens for expressions extracted from the memo in some case. I can't remember the scenario, but if you comment the similar code out in `opttoy` you can see what breaks.
</blockquote></details>

I figured as much. But it hid a bug I had (accidentally padded the children with `nil`s) and the normalization tests didn't catch that. When it becomes necessary, I'd rather have a flag or separate function which allows `nil`s.

---

*[pkg/sql/opt/index_constraints.go, line 85 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-L0BZGDbUi1rj9nvz3wq:-L0Bm8cuIWAhYrAwhzKY:b9843rc) ([raw file](https://github.com/cockroachdb/cockroach/blob/16bf712e73eb4df95ef8ce4842c61a34ddbc26c1/pkg/sql/opt/index_constraints.go#L85)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

`s/Sting/String/g`
</blockquote></details>

:musical_note: Every breath you take... :musical_note: 

---

*[pkg/sql/opt/index_constraints.go, line 189 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-L0BZxZKvhZQ9rhd0T5R:-L0BmhXRSHzv4oUAGrMC:b-zc9sxc) ([raw file](https://github.com/cockroachdb/cockroach/blob/16bf712e73eb4df95ef8ce4842c61a34ddbc26c1/pkg/sql/opt/index_constraints.go#L189)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Consider adding an enum for `direction` to enforce a `+1` or `-1` value.
</blockquote></details>

Done. Not crazy about the casts though.

---

*[pkg/sql/opt/index_constraints.go, line 331 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-L0Bedhv3Xc8Jc2IRoMb:-L0BnHSJrwKLH6LUIZrz:b-ytqr17) ([raw file](https://github.com/cockroachdb/cockroach/blob/16bf712e73eb4df95ef8ce4842c61a34ddbc26c1/pkg/sql/opt/index_constraints.go#L331)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I don't think we want to leak the contents of `private` throughout the code. Perhaps pull this out into an `isIndexedVar` method. Or should we be populating something in `scalarProps` with regards to variable indexes? I guess that might be a TODO for the future given that this code doesn't have variable indexes yet. Seems like `indexConstraintCalc` will also need a slice of indexes for the variables composing the index.
</blockquote></details>

Added a separate method. I agree we won't store this in `private`, it will be a variable index. For now the index columns are always 0, 1, 2, etc. (`splitFilter` will be used prior to calling in this code to get an expression with the right indices).

---

*[pkg/sql/opt/scalar.go, line 143 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-L0BY3EgL-VN2Lo4wS5p:-L0BoLxuWA4HEJi5apKw:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/159c322300387675b7e70767f610f05644388a38/pkg/sql/opt/scalar.go#L143)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

`s/have went/have gone/g`
</blockquote></details>

Done.

---

*[pkg/sql/opt/testdata/build-scalar, line 48 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-L0BYI1hukaHujcksCbf:-L0BofFyMyPSUucyJaXx:bfbq99i) ([raw file](https://github.com/cockroachdb/cockroach/blob/159c322300387675b7e70767f610f05644388a38/pkg/sql/opt/testdata/build-scalar#L48)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I'd like to see some other test cases here, such as normalize ORs and various combinations of ANDs and ORs.
</blockquote></details>

Good point, done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 1 of 7 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[pkg/sql/opt/index_constraints.go, line 189 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-L0BZxZKvhZQ9rhd0T5R:-L0CmNGrY251qJtxz8wi:bvuztx1) ([raw file](https://github.com/cockroachdb/cockroach/blob/16bf712e73eb4df95ef8ce4842c61a34ddbc26c1/pkg/sql/opt/index_constraints.go#L189)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Done. Not crazy about the casts though.
</blockquote></details>

Agreed, but I think this is better.

---

*[pkg/sql/opt/index_constraints.go, line 374 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-L0Cnjx3VR0f_7X8gfAG:-L0Cnjx3VR0f_7X8gfAH:b-61xlfv) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e998329f6204fc9f0d0bfcb50d4dc1a90d6c770/pkg/sql/opt/index_constraints.go#L374)):*
> ```Go
> // maximum depth). At a given depth, we create constraints on index columns
> // starting at depth and try to use the computed constraints for deeper levels
> // to extend these constraints.
> ```

I think it would be helpful to expand this comment with an example. The term `depth` is mildly confusing to me. Seems like you're calculating constraints from the first column in the index to the last. I'm also confused by the description of calculating constraints "back-to-front". Looks like `MakeIndexConstraints` starts with `depth=0`. I'm either terribly confused by these comments, or not understanding the code.

---

*[pkg/sql/opt/index_constraints.go, line 388 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-L0Cn3m0HuPdXB5WzPGs:-L0Cn3m0HuPdXB5WzPGt:b-pyl5kj) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e998329f6204fc9f0d0bfcb50d4dc1a90d6c770/pkg/sql/opt/index_constraints.go#L388)):*
> ```Go
> 	spans := LogicalSpans{MakeFullSpan()}
> 
> 	for _, e := range c.andExprs {
> ```

Would there be any simplification to sorting/grouping `andExprs` so that the children are in the same order as the indexed columns?

---

*[pkg/sql/opt/index_constraints.go, line 434 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-L0CqfLKGlBH-DqEha86:-L0CqfLKGlBH-DqEha87:btxlnqj) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e998329f6204fc9f0d0bfcb50d4dc1a90d6c770/pkg/sql/opt/index_constraints.go#L434)):*
> ```Go
> // MakeIndexConstraints generates constraints from a scalar boolean filter
> // expression. See LogicalSpans for more information on how constraints are
> // represented.
> ```

We'll need to eventually pass in a slice of indexes for the columns, right? Currently you assume the index is composed of columns `0, 1, 2, 3, ...`. Please add a TODO about this.

---

*[pkg/sql/opt/testdata/build-scalar, line 89 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-L0Cpde7vx3-TntNy8Xe:-L0Cpde7vx3-TntNy8Xf:b-fbb2hd) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e998329f6204fc9f0d0bfcb50d4dc1a90d6c770/pkg/sql/opt/testdata/build-scalar#L89)):*
> ```
>            └── inputs
>                 ├── variable (1) (type: int)
>                 └── const (10) (type: int)
> ```

Unrelated to this PR, I wonder if we should condense the formatting of `expr` a bit more. The `inputs` lines aren't adding much for scalar expressions. And we might want to list the index of children explicitly. Something like:

```
or (type: bool)
├── 0: ge (type: bool)
│   ├── variable (0) (type: int)
│   └── const (5) (type: int)
├── 1: le (type: bool)
│   ├── variable (0) (type: int)
│   └── const (10) (type: int)
├── 2: gt (type: bool)
│   ├── variable (1) (type: int)
│   └── const (0) (type: int)
└── 3: lt (type: bool)
    ├── variable (1) (type: int)
    └── const (10) (type: int)
```

---

*[pkg/sql/opt/testdata/index-constraints, line 1 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-L0CqJ01a1St9hEZa7Ec:-L0CqJ01a1St9hEZa7Ed:b10ca2m) ([raw file](https://github.com/cockroachdb/cockroach/blob/16bf712e73eb4df95ef8ce4842c61a34ddbc26c1/pkg/sql/opt/testdata/index-constraints#L1)):*
> ```
> index-constraints int
> ```

Nice tests!

---

*[pkg/sql/opt/testdata/index-constraints, line 63 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-L0CowmsyDR4iwPQRr9N:-L0CowmsyDR4iwPQRr9O:bp2yh7o) ([raw file](https://github.com/cockroachdb/cockroach/blob/16bf712e73eb4df95ef8ce4842c61a34ddbc26c1/pkg/sql/opt/testdata/index-constraints#L63)):*
> ```
> index-constraints int
> @1 > 2 AND @1 < 1
> ----
> ```

Are there no index constraints due to normalization or the constraints generation? Normalization should determine that this expression is false. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 1 of 7 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[pkg/sql/opt/index_constraints.go, line 189 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-L0BZxZKvhZQ9rhd0T5R:-L0ErmsLfnvRbnuB3VLP:bifm0kn) ([raw file](https://github.com/cockroachdb/cockroach/blob/16bf712e73eb4df95ef8ce4842c61a34ddbc26c1/pkg/sql/opt/index_constraints.go#L189)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Agreed, but I think this is better.
</blockquote></details>

Yeah, sorry, forgot to update the comment after I pulled out the cast (it was in each `return` initially).

---

*[pkg/sql/opt/index_constraints.go, line 374 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-L0Cnjx3VR0f_7X8gfAG:-L0Erz33zIRaoT8JfWV5:b-clo3as) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e998329f6204fc9f0d0bfcb50d4dc1a90d6c770/pkg/sql/opt/index_constraints.go#L374)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I think it would be helpful to expand this comment with an example. The term `depth` is mildly confusing to me. Seems like you're calculating constraints from the first column in the index to the last. I'm also confused by the description of calculating constraints "back-to-front". Looks like `MakeIndexConstraints` starts with `depth=0`. I'm either terribly confused by these comments, or not understanding the code.
</blockquote></details>

Yes, it starts with depth=0. But that calculation is finished *last*. During the computation, recursive calls first calculate constraints for i+1, i+2, etc.

If my approach is computing something back to front, I could do:
```go
for i := len(x)-1; i >= 0; i-- {
     // ** calculate something based on stuff already calculated for i+1, i+2, etc.
   }
```
or, I could do:
```go
func (i int) {
  if i is already calculated {
    return already calculated result
  }
  // * calculate something, calling recursively for i+1, i+2, as needed.
}
```

What we're doing is effectively the same - we use result for i+1,i+2 to calculate i. The first version is more straightforward, but the latter can skip computing things unnecessarily. The case I'm thinking about is a table with a lot of columns. We could have all sorts of constraints, but if none are on the first column, we shouldn't waste time with other columns.

I will add more text and an example. Do you have any other suggestions on how to make this clearer? Maybe "back to front" is not the right term to use..

---

*[pkg/sql/opt/index_constraints.go, line 388 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-L0Cn3m0HuPdXB5WzPGs:-L0EuL_hWAGRjzbcrJHB:biarjm9) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e998329f6204fc9f0d0bfcb50d4dc1a90d6c770/pkg/sql/opt/index_constraints.go#L388)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Would there be any simplification to sorting/grouping `andExprs` so that the children are in the same order as the indexed columns?
</blockquote></details>

Yeah, I will add a TODO. It will be trickier for tuples. We could also do some pre-processing and build a map of what constraints are relevant to which columns. Perhaps that could be a general helper structure - it could be reused across different indexes.

---

*[pkg/sql/opt/index_constraints.go, line 434 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-L0CqfLKGlBH-DqEha86:-L0ErIPEtPSrLgWRk0pY:b-6ei5yv) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e998329f6204fc9f0d0bfcb50d4dc1a90d6c770/pkg/sql/opt/index_constraints.go#L434)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

We'll need to eventually pass in a slice of indexes for the columns, right? Currently you assume the index is composed of columns `0, 1, 2, 3, ...`. Please add a TODO about this.
</blockquote></details>

Will add a TODO.

The sql code has to use `splitFilter` to get a filter that applies to these columns, and that function can also remap `IndexedVar`s. So we won't need it in the short term. Longer term, we want to move `splitFilter` to `opt` as well; depending if it will still have the capability to remap vars, we may need a slice.

---

*[pkg/sql/opt/testdata/build-scalar, line 89 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-L0Cpde7vx3-TntNy8Xe:-L0Euj41XqOaMGEwNn-F:b78qh4v) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e998329f6204fc9f0d0bfcb50d4dc1a90d6c770/pkg/sql/opt/testdata/build-scalar#L89)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Unrelated to this PR, I wonder if we should condense the formatting of `expr` a bit more. The `inputs` lines aren't adding much for scalar expressions. And we might want to list the index of children explicitly. Something like:

```
or (type: bool)
├── 0: ge (type: bool)
│   ├── variable (0) (type: int)
│   └── const (5) (type: int)
├── 1: le (type: bool)
│   ├── variable (0) (type: int)
│   └── const (10) (type: int)
├── 2: gt (type: bool)
│   ├── variable (1) (type: int)
│   └── const (0) (type: int)
└── 3: lt (type: bool)
    ├── variable (1) (type: int)
    └── const (10) (type: int)
```
</blockquote></details>

This looks indeed much better. Will do in a separate PR.

---

*[pkg/sql/opt/testdata/index-constraints, line 63 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-L0CowmsyDR4iwPQRr9N:-L0EupkyCY6Q2ahQlppH:b-nm8gue) ([raw file](https://github.com/cockroachdb/cockroach/blob/16bf712e73eb4df95ef8ce4842c61a34ddbc26c1/pkg/sql/opt/testdata/index-constraints#L63)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Are there no index constraints due to normalization or the constraints generation? Normalization should determine that this expression is false. 
</blockquote></details>

Good question, I'll look into it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 1 of 7 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/sql/opt/index_constraints.go, line 374 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-L0Cnjx3VR0f_7X8gfAG:-L0F7mmUElwTIbmzEH_J:bgqzxbg) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e998329f6204fc9f0d0bfcb50d4dc1a90d6c770/pkg/sql/opt/index_constraints.go#L374)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Yes, it starts with depth=0. But that calculation is finished *last*. During the computation, recursive calls first calculate constraints for i+1, i+2, etc.

If my approach is computing something back to front, I could do:
```go
for i := len(x)-1; i >= 0; i-- {
     // ** calculate something based on stuff already calculated for i+1, i+2, etc.
   }
```
or, I could do:
```go
func (i int) {
  if i is already calculated {
    return already calculated result
  }
  // * calculate something, calling recursively for i+1, i+2, as needed.
}
```

What we're doing is effectively the same - we use result for i+1,i+2 to calculate i. The first version is more straightforward, but the latter can skip computing things unnecessarily. The case I'm thinking about is a table with a lot of columns. We could have all sorts of constraints, but if none are on the first column, we shouldn't waste time with other columns.

I will add more text and an example. Do you have any other suggestions on how to make this clearer? Maybe "back to front" is not the right term to use..
</blockquote></details>

Ok, this makes more sense now. The terms `depth` and `back to front` are still confusing to me, though. While you're calculating the constraints "back to front", you are iterating from "front to back" to extend the spans. The term `depth` makes me think of a tree, and while there is a bit of tree structure here, it feels more like a list or array iteration. Unfortunately, the term `index` is overloaded here. And I don't have a better suggestion at the moment for better terminology. Perhaps just extending this comment with an example of how the spans are built up would be sufficient. 

---

*[pkg/sql/opt/index_constraints.go, line 434 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-L0CqfLKGlBH-DqEha86:-L0F70Z8vGNQKIE_EL98:bfpc40z) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e998329f6204fc9f0d0bfcb50d4dc1a90d6c770/pkg/sql/opt/index_constraints.go#L434)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Will add a TODO.

The sql code has to use `splitFilter` to get a filter that applies to these columns, and that function can also remap `IndexedVar`s. So we won't need it in the short term. Longer term, we want to move `splitFilter` to `opt` as well; depending if it will still have the capability to remap vars, we may need a slice.
</blockquote></details>

Yeah, I can see how it won't be needed in the short term. I've been imagining that the var remapping stuff will be going away in the New World Order.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655)*
<!-- Sent from Reviewable.io -->

rytaft commented:
This is a big one :). I've reviewed a few of the files and added some comments.  I'll get to the others later today.

---

Reviewed 2 of 4 files at r1, 1 of 4 files at r2, 1 of 3 files at r3.
Review status: 3 of 7 files reviewed at latest revision, 14 unresolved discussions, some commit checks failed.

---

*[pkg/sql/opt/index_constraints.go, line 29 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-L0Bk3jksZ7lxUbPdgJQ:-L0Bk3jksZ7lxUbPdgJR:b-o3d1ah) ([raw file](https://github.com/cockroachdb/cockroach/blob/16bf712e73eb4df95ef8ce4842c61a34ddbc26c1/pkg/sql/opt/index_constraints.go#L29)):*
> ```Go
> 	// Vals is a list of values on consecutive index columns.
> 	Vals tree.Datums
> 	// Inclusive is true if the boundary includes val, or false otherwise.
> ```

`val` -> `Vals`

---

*[pkg/sql/opt/index_constraints.go, line 86 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-L0FRJTMwHoc6Iwf7iMz:-L0FRJTNmL81voOdnS9y:b-innbpa) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e998329f6204fc9f0d0bfcb50d4dc1a90d6c770/pkg/sql/opt/index_constraints.go#L86)):*
> ```Go
> 
> // String formats a LogicalSpan. Inclusivity/exclusivity is shown using
> // brackets/parens. Some examples:
> ```

It would be helpful to show some examples of what it looks like when a key consists of multiple columns.

---

*[pkg/sql/opt/index_constraints.go, line 122 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-L0FS-xg6zUDVSROyNrE:-L0FS-xg6zUDVSROyNrF:b-mthlch) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e998329f6204fc9f0d0bfcb50d4dc1a90d6c770/pkg/sql/opt/index_constraints.go#L122)):*
> ```Go
> //  - no constraints: a single span with no boundaries [ - ]
> //  - a constraint on @1 > 1: a single span [/1 - ]
> //  - a constraint on @1 = 1 AND @2 > 1: a single span [/1/1 - ]
> ```

If @1 = 1 AND @2 >1, shouldn't the span be [/1/1 - /1] ?

---

*[pkg/sql/opt/index_constraints.go, line 142 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-L0FTbGS_U-zXqYQ7YGh:-L0FTbGS_U-zXqYQ7YGi:bofs88l) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e998329f6204fc9f0d0bfcb50d4dc1a90d6c770/pkg/sql/opt/index_constraints.go#L142)):*
> ```Go
> 	// If constraints[i] is nil, the constraints for that "depth" have not been
> 	// calculated yet.
> 	constraints []LogicalSpans
> ```

It would be helpful to note that `constraints` is used as the memoization data structure for the recursive function below.  When I first read the comment, I was confused and thought it was wrong since a single LogicalSpan can contain constraints on multiple columns.

---

*[pkg/sql/opt/index_constraints.go, line 285 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-L0FWvzIzWh_XJplenq6:-L0FWvzIzWh_XJplenq7:b-md2map) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e998329f6204fc9f0d0bfcb50d4dc1a90d6c770/pkg/sql/opt/index_constraints.go#L285)):*
> ```Go
> 				// The previous end key and the current start key have the same values.
> 				// Only acceptable if they are both exclusive.
> 				if ls[i-1].End.Inclusive || ls[i].Start.Inclusive {
> ```

Why can't one of them be inclusive?  e.g., `[/1 - /2), [/2 - /3)...`

---

*[pkg/sql/opt/index_constraints.go, line 301 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-L0FY2z2hmEk79tyjuuk:-L0FY2z37AE-5_iMUr6Z:by6a9nu) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e998329f6204fc9f0d0bfcb50d4dc1a90d6c770/pkg/sql/opt/index_constraints.go#L301)):*
> ```Go
> 	if !sp.Start.Inclusive {
> 		col := len(sp.Start.Vals) - 1
> 		if nextVal, hasNext := sp.Start.Vals[col].Next(c.evalCtx); hasNext {
> ```

Can you give more information about why this check allows you to convert the constraint to inclusive? It's not obvious to me...

Maybe an example would help?

---

*[pkg/sql/opt/index_constraints.go, line 321 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-L0FYU7iyvPt2UcITyRm:-L0FYU7iyvPt2UcITyRn:b-10b3dx) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e998329f6204fc9f0d0bfcb50d4dc1a90d6c770/pkg/sql/opt/index_constraints.go#L321)):*
> ```Go
> func (c *indexConstraintCalc) intersectSpan(a *LogicalSpan, b *LogicalSpan) bool {
> 	changed := false
> 	if c.compare(a.Start, b.Start, +1 /* direction */) < 0 {
> ```

Use the enum instead of `+1`.

---

*[pkg/sql/opt/index_constraints.go, line 325 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-L0FYYZzU563v7lVxbVZ:-L0FYYZzU563v7lVxbV_:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e998329f6204fc9f0d0bfcb50d4dc1a90d6c770/pkg/sql/opt/index_constraints.go#L325)):*
> ```Go
> 		changed = true
> 	}
> 	if c.compare(a.End, b.End, -1 /* direction */) > 0 {
> ```

ditto

---

*[pkg/sql/opt/index_constraints.go, line 376 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-L0Fa94vXP92lhHiOHFR:-L0Fa94vXP92lhHiOHFS:b-gk0cri) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e998329f6204fc9f0d0bfcb50d4dc1a90d6c770/pkg/sql/opt/index_constraints.go#L376)):*
> ```Go
> // to extend these constraints.
> //
> // The code uses recursivity and memoization to avoid calculating constraints
> ```

recursivity -> recursion?

---

*[pkg/sql/opt/index_constraints.go, line 407 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-L0FfpxqMWkmB0GUz2Te:-L0FfpxqMWkmB0GUz2Tf:b-sjuxpe) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e998329f6204fc9f0d0bfcb50d4dc1a90d6c770/pkg/sql/opt/index_constraints.go#L407)):*
> ```Go
> 	for i := range spans {
> 		startLen := len(spans[i].Start.Vals)
> 		if startLen > 0 && depth+startLen < len(c.types) && spans[i].Start.Inclusive {
> ```

When would startLen ever be more than 1?

---

*[pkg/sql/opt/opt_test.go, line 254 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-L0Bd2YVARUhpTBy3Yjb:-L0Bd2YVARUhpTBy3Yjc:b-nzn0z7) ([raw file](https://github.com/cockroachdb/cockroach/blob/159c322300387675b7e70767f610f05644388a38/pkg/sql/opt/opt_test.go#L254)):*
> ```Go
> 				}
> 
> 				for _, cmd := range cmds[1:] {
> ```

Why not include `build-scalar` and `index-constraints` as separate cases in this switch statement and iterate over all `cmds` in the for-loop?  It seems like it would be cleaner.  Then if you remove the `normalizeScalar` call on line 255, you can use the commands `build-scalar,normalize,index-constraints` to get the `index-constraints` functionality.

---

*[pkg/sql/opt/opt_test.go, line 318 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-L0BfnNtiK56KX8N39oQ:-L0BfnNtiK56KX8N39oR:b-dg5155) ([raw file](https://github.com/cockroachdb/cockroach/blob/16bf712e73eb4df95ef8ce4842c61a34ddbc26c1/pkg/sql/opt/opt_test.go#L318)):*
> ```Go
> func (iv *indexedVars) IndexedVarResolvedType(idx int) types.T {
> 	if idx >= len(iv.types) {
> 		panic("out of bounds IndexedVar; not enough types provided?")
> ```

Why does this end with a "?" 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-:-L0FigikR_sE5Xdiomjk:b-2sb2tu)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
TFTRs! Very helpful comments!

---

Review status: 1 of 8 files reviewed at latest revision, 14 unresolved discussions.

---

*[pkg/sql/opt/index_constraints.go, line 29 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-L0Bk3jksZ7lxUbPdgJQ:-L0FsMrTqHKvo-90ccCg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/16bf712e73eb4df95ef8ce4842c61a34ddbc26c1/pkg/sql/opt/index_constraints.go#L29)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

`val` -> `Vals`
</blockquote></details>

Done.

---

*[pkg/sql/opt/index_constraints.go, line 86 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-L0FRJTMwHoc6Iwf7iMz:-L0FsUTYVFkq72L1VVFc:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e998329f6204fc9f0d0bfcb50d4dc1a90d6c770/pkg/sql/opt/index_constraints.go#L86)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

It would be helpful to show some examples of what it looks like when a key consists of multiple columns.
</blockquote></details>

Done.

---

*[pkg/sql/opt/index_constraints.go, line 122 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-L0FS-xg6zUDVSROyNrE:-L0FsbZ9Qci2qjnsye1k:boc96a4) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e998329f6204fc9f0d0bfcb50d4dc1a90d6c770/pkg/sql/opt/index_constraints.go#L122)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

If @1 = 1 AND @2 >1, shouldn't the span be [/1/1 - /1] ?
</blockquote></details>

Good catch. Fixed it (and the one above)

---

*[pkg/sql/opt/index_constraints.go, line 142 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-L0FTbGS_U-zXqYQ7YGh:-L0Fsn4EGlrS4eIksu_E:b3wd2bp) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e998329f6204fc9f0d0bfcb50d4dc1a90d6c770/pkg/sql/opt/index_constraints.go#L142)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

It would be helpful to note that `constraints` is used as the memoization data structure for the recursive function below.  When I first read the comment, I was confused and thought it was wrong since a single LogicalSpan can contain constraints on multiple columns.
</blockquote></details>

Done. Replaced the comment entirely, the reader can just look at calcDepth now.

---

*[pkg/sql/opt/index_constraints.go, line 285 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-L0FWvzIzWh_XJplenq6:-L0FszYQiicb68v6NWrq:b-e7etxi) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e998329f6204fc9f0d0bfcb50d4dc1a90d6c770/pkg/sql/opt/index_constraints.go#L285)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Why can't one of them be inclusive?  e.g., `[/1 - /2), [/2 - /3)...`
</blockquote></details>

We don't want unnecessary breaks in spans, that span should be represented as `[/1 - /3)`

---

*[pkg/sql/opt/index_constraints.go, line 301 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-L0FY2z2hmEk79tyjuuk:-L0Ftn-WLSzeapK-EQsU:b79huwj) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e998329f6204fc9f0d0bfcb50d4dc1a90d6c770/pkg/sql/opt/index_constraints.go#L301)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Can you give more information about why this check allows you to convert the constraint to inclusive? It's not obvious to me...

Maybe an example would help?
</blockquote></details>

Added some examples to the function comment.

---

*[pkg/sql/opt/index_constraints.go, line 321 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-L0FYU7iyvPt2UcITyRm:-L0Ftxany83gERmQk41p:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e998329f6204fc9f0d0bfcb50d4dc1a90d6c770/pkg/sql/opt/index_constraints.go#L321)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Use the enum instead of `+1`.
</blockquote></details>

Done.

---

*[pkg/sql/opt/index_constraints.go, line 376 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-L0Fa94vXP92lhHiOHFR:-L0Fu1oXr1IbE9kUoQJ0:bgmaei3) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e998329f6204fc9f0d0bfcb50d4dc1a90d6c770/pkg/sql/opt/index_constraints.go#L376)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

recursivity -> recursion?
</blockquote></details>

I rewrote this entire comment, PTAL

---

*[pkg/sql/opt/index_constraints.go, line 407 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-L0FfpxqMWkmB0GUz2Te:-L0Fp9_1gxNXzxdYXsyB:bkzg74u) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e998329f6204fc9f0d0bfcb50d4dc1a90d6c770/pkg/sql/opt/index_constraints.go#L407)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

When would startLen ever be more than 1?
</blockquote></details>

Soon :) When we add support for tuple expressions e.g. `(@1, @2) >= (1, 2)`.

---

*[pkg/sql/opt/index_constraints.go, line 434 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-L0CqfLKGlBH-DqEha86:-L0FpJi0MD1mn0faGtxn:b-g2uxyt) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e998329f6204fc9f0d0bfcb50d4dc1a90d6c770/pkg/sql/opt/index_constraints.go#L434)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Yeah, I can see how it won't be needed in the short term. I've been imagining that the var remapping stuff will be going away in the New World Order.
</blockquote></details>

Ah, that makes sense, yeah we probably won't need to ever remap vars.

---

*[pkg/sql/opt/opt_test.go, line 254 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-L0Bd2YVARUhpTBy3Yjb:-L0Fri8sFHdDIuIww1mr:b-g70hwh) ([raw file](https://github.com/cockroachdb/cockroach/blob/159c322300387675b7e70767f610f05644388a38/pkg/sql/opt/opt_test.go#L254)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Why not include `build-scalar` and `index-constraints` as separate cases in this switch statement and iterate over all `cmds` in the for-loop?  It seems like it would be cleaner.  Then if you remove the `normalizeScalar` call on line 255, you can use the commands `build-scalar,normalize,index-constraints` to get the `index-constraints` functionality.
</blockquote></details>

Cool! Done.

---

*[pkg/sql/opt/opt_test.go, line 318 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-L0BfnNtiK56KX8N39oQ:-L0Frrm6OdScC5y5r_Pg:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/16bf712e73eb4df95ef8ce4842c61a34ddbc26c1/pkg/sql/opt/opt_test.go#L318)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Why does this end with a "?" 
</blockquote></details>

Fixed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-:-L0FuyXLn1-MYJtaZxp3:bhuoy58)*
<!-- Sent from Reviewable.io -->

rytaft commented:
Review status: 1 of 8 files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt/index_constraints.go, line 122 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-L0FS-xg6zUDVSROyNrE:-L0FwOVbP-1-8gwKKsZw:bsb7cki) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e998329f6204fc9f0d0bfcb50d4dc1a90d6c770/pkg/sql/opt/index_constraints.go#L122)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Good catch. Fixed it (and the one above)
</blockquote></details>

Looks good.  Just missing a `/` in the second example.

---

*[pkg/sql/opt/index_constraints.go, line 142 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-L0FTbGS_U-zXqYQ7YGh:-L0Fwb-yYELht0Jv9e0T:b2dzj1q) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e998329f6204fc9f0d0bfcb50d4dc1a90d6c770/pkg/sql/opt/index_constraints.go#L142)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Done. Replaced the comment entirely, the reader can just look at calcDepth now.
</blockquote></details>

Looks good - minor thing: `constraints[i]` -> `constraints[]`

---

*[pkg/sql/opt/index_constraints.go, line 376 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-L0Fa94vXP92lhHiOHFR:-L0G42UfXlKhQ_IDXXyJ:bttgh61) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e998329f6204fc9f0d0bfcb50d4dc1a90d6c770/pkg/sql/opt/index_constraints.go#L376)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I rewrote this entire comment, PTAL
</blockquote></details>

Very helpful!  Just fix the issues I identified inline.

---

*[pkg/sql/opt/index_constraints.go, line 407 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-L0FfpxqMWkmB0GUz2Te:-L0G4Np2KHQEfTrSk5s8:b-a9wi0j) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e998329f6204fc9f0d0bfcb50d4dc1a90d6c770/pkg/sql/opt/index_constraints.go#L407)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Soon :) When we add support for tuple expressions e.g. `(@1, @2) >= (1, 2)`.
</blockquote></details>

Got it - thanks! It might help to add that as a comment, too.

---

*[pkg/sql/opt/index_constraints.go, line 389 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-L0G3KGWwPXba0w5o8Xj:-L0G3KGWwPXba0w5o8Xk:b-xmmy46) ([raw file](https://github.com/cockroachdb/cockroach/blob/d7c0939b7aa8a653303530c2d4425922bec42bea/pkg/sql/opt/index_constraints.go#L389)):*
> ```Go
> //   - process expressions, generating span [/5 - /10]
> //   - call calcDepth(depth = 1):   // calculate constraints for @2, @3
> //     - process expressions, generating span [/2 - /2]
> ```

Do you mean `[/1 - /1]` since `@2 = 1`?

---

*[pkg/sql/opt/index_constraints.go, line 392 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-L0G3TXvUZgABr9GfQCU:-L0G3TXvUZgABr9GfQCV:bn2qdse) ([raw file](https://github.com/cockroachdb/cockroach/blob/d7c0939b7aa8a653303530c2d4425922bec42bea/pkg/sql/opt/index_constraints.go#L392)):*
> ```Go
> //     - call calcDepth(depth = 2): // calculate constraints for @3
> //       - process expressions, return span (/1 - ]
> //     - extend the keys in the span [2 - 2] with (/1 - ], return (/2/1 - /2].
> ```

`span [2 - 2]` -> `span [/1 - /1]`, `return (/2/2 - /2]` -> `return (/1/1 - /1]`

---

*[pkg/sql/opt/index_constraints.go, line 393 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-L0G3syfOnIabnUHDbFM:-L0G3syfOnIabnUHDbFN:b9o86md) ([raw file](https://github.com/cockroachdb/cockroach/blob/d7c0939b7aa8a653303530c2d4425922bec42bea/pkg/sql/opt/index_constraints.go#L393)):*
> ```Go
> //       - process expressions, return span (/1 - ]
> //     - extend the keys in the span [2 - 2] with (/1 - ], return (/2/1 - /2].
> //   - extend the keys in the span [/5 - /10] with (/2/1 - /2], return
> ```

Fix these examples too (`2` -> `1`)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 1 of 8 files reviewed at latest revision, 9 unresolved discussions.

---

*[pkg/sql/opt/index_constraints.go, line 122 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-L0FS-xg6zUDVSROyNrE:-L0G7frCnpfvO8c-IqiU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e998329f6204fc9f0d0bfcb50d4dc1a90d6c770/pkg/sql/opt/index_constraints.go#L122)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Looks good.  Just missing a `/` in the second example.
</blockquote></details>

Done.

---

*[pkg/sql/opt/index_constraints.go, line 142 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-L0FTbGS_U-zXqYQ7YGh:-L0G7jusuvPK1MvNAmAi:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e998329f6204fc9f0d0bfcb50d4dc1a90d6c770/pkg/sql/opt/index_constraints.go#L142)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Looks good - minor thing: `constraints[i]` -> `constraints[]`
</blockquote></details>

Done.

---

*[pkg/sql/opt/index_constraints.go, line 407 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-L0FfpxqMWkmB0GUz2Te:-L0G8BcBuutnQDWb0-cq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e998329f6204fc9f0d0bfcb50d4dc1a90d6c770/pkg/sql/opt/index_constraints.go#L407)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Got it - thanks! It might help to add that as a comment, too.
</blockquote></details>

Done.

---

*[pkg/sql/opt/index_constraints.go, line 389 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-L0G3KGWwPXba0w5o8Xj:-L0G7n1xWUnYTdhEWqFK:boxv3fm) ([raw file](https://github.com/cockroachdb/cockroach/blob/d7c0939b7aa8a653303530c2d4425922bec42bea/pkg/sql/opt/index_constraints.go#L389)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Do you mean `[/1 - /1]` since `@2 = 1`?
</blockquote></details>

Changed the expression to  `@2 = 2`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655)*
<!-- Sent from Reviewable.io -->

rytaft commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>

---

Reviewed 1 of 4 files at r2, 2 of 3 files at r3, 4 of 4 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/sql/opt/scalar.go, line 172 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-L0GAM9quIaRt2jNgDaD:-L0GAM9quIaRt2jNgDaE:bm8hrpq) ([raw file](https://github.com/cockroachdb/cockroach/blob/58a225721ad35e0ea55312a6795d125ff7740e7a/pkg/sql/opt/scalar.go#L172)):*
> ```Go
> 		if found {
> 			saved := e.children
> 			e.children = make([]*expr, 0, numNewChildren)
> ```

shouldn't the capacity be `numNewChildren + len(e.children)`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-:-L0GCoLn0LrMBABOiHZg:b-jbsf8z)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 7 of 8 files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/sql/opt/scalar.go, line 172 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-L0GAM9quIaRt2jNgDaD:-L0GFUWHUFTdXpBkc6a8:b-14dhc) ([raw file](https://github.com/cockroachdb/cockroach/blob/58a225721ad35e0ea55312a6795d125ff7740e7a/pkg/sql/opt/scalar.go#L172)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

shouldn't the capacity be `numNewChildren + len(e.children)`?
</blockquote></details>

numNewChildren is not the best name, it's the new number of children. Renamed to newNumChildren

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655)*
<!-- Sent from Reviewable.io -->

rytaft commented:
Reviewed 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt/scalar.go, line 172 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-L0GAM9quIaRt2jNgDaD:-L0GNXwgLLAU7P6gA47j:bjxptv) ([raw file](https://github.com/cockroachdb/cockroach/blob/58a225721ad35e0ea55312a6795d125ff7740e7a/pkg/sql/opt/scalar.go#L172)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

numNewChildren is not the best name, it's the new number of children. Renamed to newNumChildren
</blockquote></details>

Ah sorry I missed that you already included `len(e.children)`.  Anyway, I think the new name is good - thanks!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
*[pkg/sql/opt/testdata/index-constraints, line 63 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-L0CowmsyDR4iwPQRr9N:-L0GXBpcFrBV7g9xQi65:bg6ho71) ([raw file](https://github.com/cockroachdb/cockroach/blob/16bf712e73eb4df95ef8ce4842c61a34ddbc26c1/pkg/sql/opt/testdata/index-constraints#L63)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Good question, I'll look into it.
</blockquote></details>

It doesn't look like it's getting normalized, the filter coming into `MakeIndexConstraints` is
```
filter: and (type: bool)
 └── inputs
      ├── gt (type: bool)
      │    └── inputs
      │         ├── variable (0) (type: int)
      │         └── const (2) (type: int)
      └── lt (type: bool)
           └── inputs
                ├── variable (0) (type: int)
                └── const (1) (type: int)
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 1 of 8 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt/testdata/index-constraints, line 63 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-L0CowmsyDR4iwPQRr9N:-L0GYreJ8ihKwYT9VIkP:baw7y31) ([raw file](https://github.com/cockroachdb/cockroach/blob/16bf712e73eb4df95ef8ce4842c61a34ddbc26c1/pkg/sql/opt/testdata/index-constraints#L63)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

It doesn't look like it's getting normalized, the filter coming into `MakeIndexConstraints` is
```
filter: and (type: bool)
 └── inputs
      ├── gt (type: bool)
      │    └── inputs
      │         ├── variable (0) (type: int)
      │         └── const (2) (type: int)
      └── lt (type: bool)
           └── inputs
                ├── variable (0) (type: int)
                └── const (1) (type: int)
```
</blockquote></details>

We don't call `tree.NormalizeExpr`. I will add that (in another PR). Not sure if it should be part of `normalize` directive or we should have a separate one.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 1 of 8 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt/index_constraints.go, line 394 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-L0Gb-6zhOT_KB4Al8Eh:-L0Gb-6zhOT_KB4Al8Ei:b-7wcog7) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8013303932a8b0d02fbcfce19002c9185107f3d/pkg/sql/opt/index_constraints.go#L394)):*
> ```Go
> //     - extend the keys in the span [/2 - /2] with (/1 - ], return (/2/1 - /2].
> //   - extend the keys in the span [/5 - /10] with (/2/1 - /2], return
> //     (/5/2/1 - /10/2].
> ```

This comment is much nicer. Thanks.

---

*[pkg/sql/opt/testdata/index-constraints, line 63 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-L0CowmsyDR4iwPQRr9N:-L0Gb78FSTZx4RSP9O_G:b211l4o) ([raw file](https://github.com/cockroachdb/cockroach/blob/16bf712e73eb4df95ef8ce4842c61a34ddbc26c1/pkg/sql/opt/testdata/index-constraints#L63)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

We don't call `tree.NormalizeExpr`. I will add that (in another PR). Not sure if it should be part of `normalize` directive or we should have a separate one.
</blockquote></details>

When we generate index constraints for a logically false expression, is there a signal that a table scan is not necessary?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655)*
<!-- Sent from Reviewable.io -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 1 of 8 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-:-L0GbSAPWV3rML6FnOCv:bnfp4nl)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 1 of 8 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt/testdata/index-constraints, line 63 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655#-L0CowmsyDR4iwPQRr9N:-L0GdAj6MBEeuDoOXtWL:b-3x3t5m) ([raw file](https://github.com/cockroachdb/cockroach/blob/16bf712e73eb4df95ef8ce4842c61a34ddbc26c1/pkg/sql/opt/testdata/index-constraints#L63)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

When we generate index constraints for a logically false expression, is there a signal that a table scan is not necessary?
</blockquote></details>

The signal is no spans, i.e. empty `LogicalSpans` in this code, empty `[]roachpb.Span`s in the existing code where it causes the `scanNode` to be converted to a `zeroNode`: https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/opt_index_selection.go#L229

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20655)*
<!-- Sent from Reviewable.io -->
